mlos_core/mlos_core/optimizers/bayesian_optimizers/smac_optimizer.py

```diff
@@ -0,0 +1,212 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Contains the wrapper class for SMAC Bayesian optimizers.
+"""
+
+from pathlib import Path
+from typing import Mapping, Optional
+
+import ConfigSpace
+import numpy.typing as npt
+import pandas as pd
+
+from mlos_core.optimizers.bayesian_optimizers import BaseBayesianOptimizer
+from mlos_core.spaces.adapters.adapter import BaseSpaceAdapter
+
+
+class SmacOptimizer(BaseBayesianOptimizer):
+    """Wrapper class for SMAC based Bayesian optimization.
+
+    Parameters
+    ----------
+    parameter_space : ConfigSpace.ConfigurationSpace
+        The parameter space to optimize.
+
+    space_adapter : BaseSpaceAdapter
+        The space adapter class to employ for parameter space transformations.
+
+    seed : Optional[int]
+        Seed is used to make results reproducible. A value of `None` will use a random seed.
+
+    run_name : str
+        Name of this run. This is used to easily distinguish across different runs.
+
+    output_directory : str
+        The directory where SMAC output will saved.
+
+    n_random_init : Optional[int]
+        Number of points evaluated at start to bootstrap the optimizer.
+
+    n_random_probability: Optional[float]
+        Probability of choosing to evaluate a random configuration during optimization.
+        Setting this to a higher value favors exploration over exploitation.
+
+    n_workers: int
+        Number of parallel workers to use.
+    """
+
+    def __init__(  # pylint: disable=too-many-arguments,too-many-locals
+        self,
+        parameter_space: ConfigSpace.ConfigurationSpace,
+        space_adapter: Optional[BaseSpaceAdapter] = None,
+        seed: Optional[int] = None,
+        run_name: str = 'smac',
+        output_directory: str = 'smac3_output',
+        n_random_init: Optional[int] = 10,
+        n_random_probability: Optional[float] = 0.1,
+        n_workers: int = 1,
+    ):
+        super().__init__(parameter_space, space_adapter)
+
+        # pylint: disable=import-outside-toplevel
+        from smac import HyperparameterOptimizationFacade as Optimizer_Smac
+        from smac import Scenario
+        from smac.intensifier.abstract_intensifier import AbstractIntensifier
+        from smac.initial_design import LatinHypercubeInitialDesign
+        from smac.main.config_selector import ConfigSelector
+        from smac.random_design.probability_design import ProbabilityRandomDesign
+        from smac.runhistory import TrialInfo
+
+        self.trial_info_map: Mapping[ConfigSpace.Configuration, TrialInfo] = {}  # Stores TrialInfo instances returned by .ask()
+
+        # Instantiate Scenario
+        output_directory = Path(output_directory)
```

Comment:
Is an output directory needed?
If not, we may want to default to not-storing.

Reply:
Alt, when this gets instantiated via `mlos_bench` we could consider to use the `tempdir_context` to store the intermediaries in a tempdir that gets automatically cleaned up.

Reply:
Unfortuantely, SMAC requires this to be valid. I'll use Python's `TemporaryDirectory` instance for now. Will switch to `tempdir_context` when integrating to `mlos_bench`.